packages/flutter_tools/lib/src/custom_devices/custom_device.dart

```diff
@@ -0,0 +1,724 @@
+
+  @override
+  Future<int> forward(int devicePort, {int hostPort}) async {
+    int actualHostPort = (hostPort == 0 || hostPort == null) ? devicePort : hostPort;
+    int tries = 0;
+
+    while ((numTries == null) || (tries < numTries)) {
+      // when the desired host port is already forwarded by this Forwarder,
+      // choose another one
+      while (_forwardedPorts.any((ForwardedPort port) => port.hostPort == actualHostPort)) {
+        actualHostPort++;
+      }
+
+      final ForwardedPort port = await _tryForward(devicePort, actualHostPort);
+
+      if (port != null) {
+        _forwardedPorts.add(port);
+        return actualHostPort;
+      } else {
+        // null value means the forwarding failed (for whatever reason)
+        // increase port by one and try again
+        actualHostPort++;
+        tries++;
+      }
+    }
+
+    throw ToolExit('Forwarding port for custom device $_deviceName failed after $tries tries.');
+  }
+
+  @override
+  List<ForwardedPort> get forwardedPorts => List<ForwardedPort>.unmodifiable(_forwardedPorts);
+
+  @override
+  Future<void> unforward(ForwardedPort forwardedPort) async {
+    assert(_forwardedPorts.contains(forwardedPort));
+
+    // since a forwarded port represents a running process launched with
+    // the forwardPortCommand, unforwarding is as easy as killing the proces
+    _processManager.killPid(forwardedPort.context.pid);
+    _forwardedPorts.remove(forwardedPort);
+  }
+}
+
+class CustomDeviceAppSession {
+  CustomDeviceAppSession({
+    @required this.name,
+    @required CustomDevice device,
+    @required ApplicationPackage appPackage,
+    @required Logger logger,
+    @required ProcessManager processManager
+  }) : _appPackage = appPackage,
+       _device = device,
+       _logger = logger,
+       _processManager = processManager,
+       logReader = CustomDeviceLogReader(name);
+
+  final String name;
+  final CustomDevice _device;
+  final ApplicationPackage _appPackage;
+  final Logger _logger;
+  final ProcessManager _processManager;
+  final CustomDeviceLogReader logReader;
+
+  Process _process;
+
+  Future<LaunchResult> start({
+    String mainPath,
+    String route,
+    DebuggingOptions debuggingOptions,
+    Map<String, dynamic> platformArgs,
+    bool prebuiltApplication = false,
+    bool ipv6 = false,
+    String userIdentifier
+  }) async {
+    final List<String> interpolated = interpolateCommand(
+      _device._config.runDebugCommand,
+      <String, String>{
+        'remotePath': '/tmp/',
+        'appName': _appPackage.name
+      }
+    );
+
+    final Process process = await _processManager.start(interpolated);
+
+    assert(_process == null);
+    _process = process;
+
+    final ProtocolDiscovery discovery = ProtocolDiscovery.observatory(
+      logReader,
+      portForwarder: _device._config.usesPortForwarding ? _device.portForwarder : null,
+      hostPort: null,
+      devicePort: null,
+      logger: _logger,
+      ipv6: ipv6,
+    );
+
```

Comment:
uber-nit: the spacing here is a little odd, could you collect the related bits together?